tensor2tensor/layers/common_attention.py

```
@@ -1465,7 +1479,12 @@ def dot_product_attention(q,
     if bias is not None:
       bias = common_layers.cast_like(bias, logits)
       logits += bias
-    weights = tf.nn.softmax(logits, name="attention_weights")
+    # TODO: pass hparams through so we can call get_tf_activation_dtype?
+    if logits.dtype == tf.float16:
```

can probably make this block more elegant

---

something like

```
def maybe_upcast(logits, hparams):
  if hparams.activation_dtype == tf.float16 and hparams.weights_dtype == tf.float32:
    return tf.cast(logits, tf.float32)


logits = maybe_upcast(logits)
weights = tf.nn.softmax(logits, name="attention_weights")
weights = tf.cast(weights, get_tf_activation_dtype(...))
````

---

Right this is definitely cleaner. In the same spirit, does it make sense to just create a softmax function that internally optionally upcasts and then returns the same type?
e.g
```
def mp_softmax(logits, *args, **kwargs):
    return_type = logits.dtype
    logits = maybe_upcast(logits)
    return tf.cast(tf.nn.softmax(logits, *args, **kwargs), out_type)
...
weights = mp_softmax(logits, name="attention_weights")
```
It seems like almost every softmax fn would need the fp32 upcast given the nvidia docs, so this might be cleaner in the long run.

---

Also, hparams aren't passed in to dot_product_attention. So we'll have to find another way to determine whether to upcast. Thinking on this

---

A thought: If we make the mp_sofmax function, that might help us get around the maybe_upcast function. Since in the context of the softmax we can assume we want to upcast regardless of the hparam context. Thoughts?

---

I just put up 
my v1 of the change here: https://github.com/medicode/tensor2tensor/pull/112/commits/14308dd646a605dee895433d1b720ee231bfcc12
I tested that it works by printing the value of activation_dtype in dot_product_attention and it was always fp16.

Design notes and questions:
1. I opted to pass in the activation_dtype and weight_dtype explicitly instead of hparams, because that seemed more consistent with the existing code. Let me know if that is a bad call due to some later design impact
2. I'm confused about the behaviour of maybe_upcast. Why do we only upcast if weights_dtype is fp32? What behvaiour would we expect if you set weights_dtype to bfloat16 for example? More broadly, should we just throw an error if we see some pairing of activation_dtype + weight_dtype that isn't fp16 + fp32?

---

Resolving this after our slack convo